pilot/pkg/config/event/store.go

```diff
@@ -0,0 +1,229 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package event
+
+import (
+	"errors"
+	"fmt"
+
+	"istio.io/pkg/ledger"
+	"istio.io/pkg/log"
+
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/config/event"
+	"istio.io/istio/pkg/config/schema/collection"
+	"istio.io/istio/pkg/config/schema/collections"
+	resourceSchema "istio.io/istio/pkg/config/schema/resource"
+)
+
+var (
+	errUnsupported        = errors.New("this operation is not supported by event-based config store")
+	supportedSchemas      = collections.Pilot
+	supportedSchemasArray = supportedSchemas.All()
+	serviceEntryKind      = collections.IstioNetworkingV1Alpha3Serviceentries.Resource().GroupVersionKind()
+)
+
+const (
+	ledgerLogf = "error tracking pilot config versions for mcp distribution: %v"
+)
+
+// Options stores the configurable attributes of a Control
+type Options struct {
+	ClusterID    string
+	DomainSuffix string
+	XDSUpdater   model.XDSUpdater
+	ConfigLedger ledger.Ledger
+}
+
+// store is a model.ConfigStoreCache populated by events.
+type store struct {
+	serviceEntryHandlers []func(model.Config, model.Config, model.Event)
+	ledger               ledger.Ledger
+	byGVK                map[resourceSchema.GroupVersionKind]*collectionStore
+	byCollection         map[collection.Name]*collectionStore
+	xdsUpdater           model.XDSUpdater
+}
+
+// NewConfigStoreAndHandler creates a new model.ConfigStoreCache and an event.Handler used to populate it.
+func NewConfigStoreAndHandler(options *Options) (model.ConfigStoreCache, event.Handler) {
+	s := &store{
+		serviceEntryHandlers: make([]func(model.Config, model.Config, model.Event), 0),
+		ledger:               options.ConfigLedger,
+		byGVK:                make(map[resourceSchema.GroupVersionKind]*collectionStore, len(supportedSchemasArray)),
+		byCollection:         make(map[collection.Name]*collectionStore, len(supportedSchemasArray)),
+		xdsUpdater:           options.XDSUpdater,
+	}
+
+	// Create the collection stores.
+	for _, schema := range supportedSchemasArray {
+		cstore := newCollectionStore(schema, options.DomainSuffix, s.onEvent)
+		s.byGVK[schema.Resource().GroupVersionKind()] = cstore
+		s.byCollection[schema.Name()] = cstore
+	}
+
+	return s, s
+}
+
+func (s *store) onEvent(schema collection.Schema, eventKind event.Kind, prev, conf *model.Config) {
+	// Update the ledger.
+	ledgerKey := model.Key(schema.Resource().Kind(), conf.Namespace, conf.Name)
+	switch eventKind {
+	case event.Added, event.Updated:
+		_, err := s.ledger.Put(ledgerKey, conf.ResourceVersion)
+		if err != nil {
+			log.Warnf(ledgerLogf, err)
+		}
+	case event.Deleted:
+		err := s.ledger.Delete(ledgerKey)
+		if err != nil {
+			log.Warnf(ledgerLogf, err)
+
+		}
+	}
+
+	gvk := schema.Resource().GroupVersionKind()
+	if gvk == serviceEntryKind {
+		s.dispatchServiceEntry(prev, conf, convertEventKind(eventKind))
+	} else {
+		s.xdsUpdater.ConfigUpdate(&model.PushRequest{
```

Comment:
The old MCP controller always ignore incremental MCP configs (always sent full update), should this event store push updates according to the incremental flag on the config?